### PR TITLE
Make entity with Start / Stop functions idempotent

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Make entity with Start / Stop functions idempotent
* Remove misleading / unnecessary dynamic configs related to NDC
* Minor issue fixes

NOTE: 
* when running in `kafka` replication mode, both tx and rx will use Kafka replication stack
* when running in `kafka-to-rpc` replication mode
  * tx will use RPC stack
  * rx will use both RPC stack and Kafka stack
* when running in `rpc` replication mode, both tx and rx will use RPC replication stack

**Why?**
Code logic should guarantee at most once start / stop

**How did you test it?**
Run locally with NDC w/ Kafka & NDC w/ RPC, NDC with Kafka -> RPC needs to be tested by larger cluster 

**Potential risks**
N/A

### CODE REVIEW

File: service/history/handler.go [hunk 1]
```diff
@@ -124,11 +126,21 @@ func NewHandler(
...
 	if h.GetClusterMetadata().IsGlobalNamespaceEnabled() {
-		var err error
-		h.publisher, err = h.GetMessagingClient().NewProducerWithClusterName(h.GetClusterMetadata().GetCurrentClusterName())
-		if err != nil {
-			h.GetLogger().Fatal("Creating kafka producer failed", tag.Error(err))
+		if h.GetClusterMetadata().GetReplicationConsumerConfig().Type == serviceConfig.ReplicationConsumerTypeKafka {
+			var err error
+			h.publisher, err = h.GetMessagingClient().NewProducerWithClusterName(h.GetClusterMetadata().GetCurrentClusterName())
+			if err != nil {
+				h.GetLogger().Fatal("Creating kafka producer failed", tag.Error(err))
```
Comment (samarabbas) [inline service/history/handler.go:142]: I guess it does not matter anymore as we are getting rid of Kafka, but in general preventing history service to come up if Kafka is down is a bad design choice.

File: service/history/replicationTaskProcessor.go [hunk 2]
```diff
@@ -210,11 +210,9 @@ func (p *ReplicationTaskProcessorImpl) eventLoop() {
 			))
 
 		case <-cleanupTimer.C:
-			if p.config.EnableCleanupReplicationTask() {
-				if err := p.cleanupReplicationTasks(); err != nil {
-					p.logger.Error("Failed to clean up replication messages.", tag.Error(err))
-					p.metricsClient.Scope(metrics.ReplicationTaskCleanupScope).IncCounter(metrics.ReplicationTaskCleanupFailure)
-				}
+			if err := p.cleanupReplicationTasks(); err != nil {
```
Comment (samarabbas) [inline service/history/replicationTaskProcessor.go:213]: Looks like this was previously guarded by `EnableCleanupReplicationTask` flag.  Now looks like this is always needed.  Just want to confirm the intention is if `GlobalNamespace` is enabled then we want run cleanupReplicationTasks in all cases?

File: service/history/replicatorQueueProcessor.go [hunk 3]
```diff
@@ -304,6 +304,13 @@ func (p *replicatorQueueProcessorImpl) getTasks(
 
 	if lastReadTaskID == persistence.EmptyQueueMessageID {
 		lastReadTaskID = p.shard.GetClusterReplicationLevel(pollingCluster)
+	} else {
+		if err := p.shard.UpdateClusterReplicationLevel(
```
Comment (samarabbas) [inline service/history/replicatorQueueProcessor.go:308]: Do we really want to `UpdateClusterReplicationLevel` on every read?

Comment (samarabbas) [reply service/history/replicatorQueueProcessor.go:308]: Looking at the logic looks like we were doing it for every replication task which was even worse.  Let's have a quick chat tomorrow to see if we can move it to a back ground task and do it even less frequently then how it is coded up right now.

File: service/history/service.go [hunk 4]
```diff
@@ -184,7 +184,6 @@ func (s *Service) Stop() {
 	remainingTime = s.sleep(shardOwnershipTransferDelay, remainingTime)
 
 	s.GetLogger().Info("ShutdownHandler: No longer taking rpc requests")
-	s.handler.PrepareToStop()
```
Comment (samarabbas) [inline service/history/service.go:187]: Why?  I thought this is for graceful shutdown.

File: service/history/handler.go (hunk 1 above)
Comment (wxing1292) [reply service/history/handler.go:142]: yep

File: service/history/replicationTaskProcessor.go (hunk 2 above)
Comment (wxing1292) [reply service/history/replicationTaskProcessor.go:213]: replicationTaskProcessor will only be initialize when there are other clusters to poll from

File: service/frontend/service.go [hunk 5]
```diff
@@ -256,9 +251,7 @@ func (s *Service) Start() {
 	clusterMetadata := s.GetClusterMetadata()
 	if clusterMetadata.IsGlobalNamespaceEnabled() {
 		consumerConfig := clusterMetadata.GetReplicationConsumerConfig()
-		if consumerConfig != nil &&
-			consumerConfig.Type == config.ReplicationConsumerTypeRPC &&
-			s.config.EnableRPCReplication() {
+		if consumerConfig.Type == config.ReplicationConsumerTypeRPC || consumerConfig.Type == config.ReplicationConsumerTypeKafkaToRPC {
```
Comment (samarabbas) [inline service/frontend/service.go:254]: I see this condition is repeated in many places.  Might be useful to create a helper method.

File: service/history/replicatorQueueProcessor.go (hunk 3 above)
Comment (wxing1292) [reply service/history/replicatorQueueProcessor.go:308]: https://github.com/temporalio/temporal/blob/622852f5c1abaae5e1f966fc532d5f74660c170c/service/history/shard/context_impl.go#L895

there is check preventing too many updates to DB
in mem change is fine

File: service/history/service.go (hunk 4 above)
Comment (wxing1292) [reply service/history/service.go:187]: the status is already updated: https://github.com/temporalio/temporal/pull/1037/files/5ca23ff1a2b3f42be78d69b590b55644feebf087#diff-162efb4a3898151e1b88f08e0caacacc43a8212abda815891ac443069b534852R154

File: service/frontend/service.go (hunk 5 above)
Comment (wxing1292) [reply service/frontend/service.go:254]: this check will only exists for one version, i.e. the version which RPC stack is ready, then kafka will be removed in the next version and all `config.ReplicationConsumerType*` will be removed

File: service/history/replicatorQueueProcessor.go (hunk 3 above)
Comment (wxing1292) [reply service/history/replicatorQueueProcessor.go:308]: this getTasks will only be called when target cluster calling source cluster

Comment (wxing1292) [reply service/history/replicatorQueueProcessor.go:308]: > Looking at the logic looks like we were doing it for every replication task which was even worse.
not each replication task, but each get tasks call

File: service/history/replicationTaskProcessor.go (hunk 2 above)
Comment (samarabbas) [reply service/history/replicationTaskProcessor.go:213]: Might be worth mentioning this as a comment here.

**Review by samarabbas — APPROVED** [review]

**Pull request merged**